### ISSUE TITLE
Add relative and absolute tolerances for matrix_rank, pinv

### ISSUE THREAD

**Issue by IvanYashchuk** [pytorch/pytorch#54151]
## 🚀 Feature

Both `torch.linalg.matrix_rank` and `torch.linalg.pinv` calculate singular values of the provided matrix and truncate them based on the specified tolerance (argument is called `rcond` for `torch.linalg.pinv` and `tol` for `torch.linalg.matrix_rank`).
Currently implemented behavior for setting the tolerance and the default tolerance values follow NumPy.
However, NumPy is not consistent in the default values and in treating the provided tolerance as relative or absolute.

|             | default<br>for the argument | default<br>tolerance  | truncation criteria<br>(if default) | truncation criteria<br>(if specified) |
|-------------|-----------------------------|-----------------------|-------------------------------------|---------------------------------------|
| matrix_rank | None                        | eps * max(rows, cols) | tol * max(singular_values)             | tol                                   |
| pinv        | 1e-15                       | 1e-15                 | tol * max(singular_values)             | tol * max(singular_values)               |

The proposal is to implement a unified way to specify the absolute or relative tolerances for the truncation of singular values as following:
```python
def matrix_rank_or_pinv(input, *, atol = 0, rtol = default_rtol):
    ...
    singular_values = ... # compute singular values of input
    truncation_criteria = max(atol, rtol * max(singular_values) )
    truncated_singular_values = singular_values > truncation_criteria
    ...
```
Possible choices of `default_rtol`:

- NumPy uses `eps * max(rows, cols)` for `matrix_rank` and `1e-15` for `pinv`
- TensorFlow uses the same default as NumPy for [`matrix_rank`](https://www.tensorflow.org/api_docs/python/tf/linalg/matrix_rank) but `10 * eps * max(rows, cols)` for [`pinv`](https://www.tensorflow.org/api_docs/python/tf/linalg/pinv)
- JAX uses the same defaults as TensorFlow
- Julia uses `eps * min(rows, cols)` both for [`pinv`](https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/#LinearAlgebra.pinv) and [`matrix_rank`](https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/#LinearAlgebra.rank)

Use of `max(atol, rtol * ...)` for defining the truncation criteria follows [math.isclose](https://docs.python.org/3/library/math.html#math.isclose).

Backwards compatibility / NumPy compatibility:
```python
def matrix_rank(input, tol = None):
    if tol is None:
        return matrix_rank(input, atol = 0, rtol = eps * max(rows, cols))
    else:
        return matrix_rank(input, atol = tol, rtol = 0)

def pinv(input, rcond = 1e-15)
    return pinv(input, atol = 0, rtol = rcond)
```

**Comment by vadimkantorov** [comment]
Things are probably going to be worse for fp16 or other short bitwidth types https://github.com/pytorch/pytorch/issues/35666. So at least thresholds/epsilons should be configurable...

**Comment by mruberry** [comment]
Just to make sure I understand this issue:

- NumPy's matrix_rank() and pinv() have differently named kwargs that control values below which singular values are set to zero (my understanding is they are not "truncated" but set to zero)
- the behavior of these kwargs is wacky
- it would make more sense to use atol, which if set would specify a number to test against, and rtol, which if set would specify a number to be computed using the maximal singular value
- if both atol and rtol are set, then any singular value which failed either criteria would be set to zero
- the historic rcond and tol kwargs could be supported just as they are, but we could encourage users to use atol and rtol instead

Does that sound right, @IvanYashchuk?

**Comment by IvanYashchuk** [comment]
That's a great summary! All the points are correct.

**Comment by vadimkantorov** [comment]
If default cut-offs are for fp32 and do not adapt automagically to fp16 inputs, then probably docs should have suggestions for cut-offs for fp16 inputs and clearly indicate that it's recommended that the user calls the function with the cut-off corresponding to dtype

**Comment by IvanYashchuk** [comment]
fp16 inputs are not supported for CPU and CUDA and I doubt they will ever be in near future. If it was supported it wouldn't be an issue: `eps` in the first message stands for "machine precision", the actual value for specified dtype can be obtained with `torch.finfo(torch.float16).eps`, so the default relative tolerance would adapt to dtype of input.

**Comment by mruberry** [comment]
This sounds like a good UX change to me. We'll have to be a little careful in the docs how we promote atol/rtol over the historic arguments and how we clarify the default behavior.

**Comment by Lezcano** [comment]
Summary of offline discussion. All this applies to `pinv`, `matrix_rank` and `lstsq`.

- The name in the documentation will be `rtol`, but we will also support `rcond`. The support of `rcond` will not be documented.
- We will add `atol` (and silently support `acond`) for `linalg.matrix_rank`, as it is the only one for which we can do this.
- The default for `rtol` will be `eps*max(rows, cols)` for all the three operations to conform with the array API. When doing this, we'll be able to remove the warnings in these operations that say that "we may change the default `rtol` value in the future".

**Comment by nikitaved** [comment]
The current default values for tolerances are too conservative for larger matrices of floating type, for example:

```python
In [25]: a = torch.rand(1, 1, 1024, 1024).cuda()

In [26]: a[..., -1, :] = 0

In [27]: a[..., :, -1] = 0

In [28]: torch.linalg.matrix_rank(a)
Out[28]: tensor([[1019]], device='cuda:0')

In [29]: a.svd()[1].abs().topk(k=5, largest=False)
Out[29]: 
torch.return_types.topk(
values=tensor([[[0.0000, 0.0011, 0.0157, 0.0202, 0.0401]]], device='cuda:0'),
indices=tensor([[[1023, 1022, 1021, 1020, 1019]]], device='cuda:0'))

```
```

**Comment by IvanYashchuk** [comment]
`torch.rand` is samples from uniform distribution, not Gaussian. It's likely not full rank even before you add zeros on the diagonal.

The default relative tolerance for `matrix_rank` (`eps*max(rows, cols)`) is common and won't be changed for something else because it's not an easy choice to decide the truncation criteria. Therefore in this proposal and [in the implementation](https://github.com/pytorch/pytorch/pull/63102), it's possible to choose between absolute tolerance that doesn't depend on singular values and relative tolerance. Maybe useful reading is section 6. How small is small from https://doi.org/10.1137/0905030? There and in some other references it's suggested to use the truncation criteria as `eps` times some growing function of the number of rows and columns.

> A side note, lstsq uses the threshold based on the largest singular value irrespective of the size of the input.

This is not the case according to the documentation:

> If rcond=None, rcond is set to the machine precision of the dtype of A times max(m, n)

**Comment by Lezcano** [comment]
The current default tolerance for `matrix_rank` is `\sigma_0 * max(m, n) * eps`. This is what's causing this weird behaviour. When this issue is solved, it'll be `max(m,n)*eps` which will return the correct results for these random matrices.

The default values for `matrix_rank` proposed in this issue start being problematic for matrices of shape 10k x 10k or 100k x 100k. If you are using such matrices in single precision in your code, you'll need to be very careful of how any linalg operation handles them regardless, so I don't think that a note in the docs is necessary.

About using SVD, well, matrix_rank uses `svdvals` behind the scenes, so I'm afraid you won't be able to do much better on your own using SVD than using the current `matrix_rank` with a reasonable atol / rtol (when this issue is implemented, of course).

As a side note, I don't think that a matrix in R^{1024 x 1024} distributed as U(0,1) will have a singular value close to zero :)

**Comment by nikitaved** [comment]
> `torch.rand` is samples from uniform distribution, not Gaussian. It's likely not full rank even before you add zeros on the diagonal.

I agree, fixed to `randn`. Nevertheless, I disagree it is not full-rank, which you can see from the analysis of smallest singular values. And generating a rank-deficient matrix from random procedures should be hard as these matrices are measure zero in the probability space.

> > A side note, lstsq uses the threshold based on the largest singular value irrespective of the size of the input.
> 
> This is not the case according to the documentation:
> 
> > If rcond=None, rcond is set to the machine precision of the dtype of A times max(m, n)

It is, but it is different for different drivers. For example, if `gelss` is used, `rcond` becomes `rtol` which is explained in:
http://www.netlib.org/lapack/explore-html/d7/d3b/group__double_g_esolve_gaa6ed601d0622edcecb90de08d7a218ec.html

**Comment by nikitaved** [comment]
> About using SVD, well, matrix_rank uses `svdvals` behind the scenes, so I'm afraid you won't be able to do much better on your own using SVD than using the current `matrix_rank` with a reasonable atol / rtol (when this issue is implemented, of course).

Since it uses SVD, I might skip on trying understand what these atol/rtol are and just implement my own custom thresholding function. Some LAPACK methods, by the way, allow to pass a custom thresholding function.

**Comment by Lezcano** [comment]
For `lstsq` all drivers have a similar behaviour when it comes to `rtol`, and that is, they are truncated whenever they are smaller than `\sigma_0 * rtol`. For why this is also the case for all the drivers see https://github.com/data-apis/array-api/issues/216

While the behaviour is certainly not consistent as one multiplies the `rtol` you give it by `sigma_0` and the other one uses its value raw, the defaults are not far from each other. For example, for square gaussian random matrices, we have that that the largest eigenvalue behaves as O(\sqrt(n)). This disonance of O(\sqrt(n)) vs n not great but it's not terrible. In practice, it feels like its something like \sigma_0 \approx sqrt(n) / 2. As such, mathematically, perhaps a better default would be `sqrt(m+n)/2*eps`, but it starts getting a bit too technical perhaps? Also, there's the consistency with the other libraries and all that...
